### PR TITLE
chore(modules/engine): deal with adapters and state at the output

### PR DESCRIPTION
**Pull request by maxvektor**
All adapters and state logic now lives at the native output. The same should work with other outputs in future.
screen shot 2019-03-01 at 4 43 22 pm

Before merge:
- [x] remove all adapters logic from playback-engine
- [x] remove all play/pause logic from playback-engine
- [x] there is some wierd calls to `/null`. will investigate and fix it

### CODE REVIEW

File: src/modules/playback-engine/playback-engine.ts [hunk 1]
```diff
@@ -662,24 +635,22 @@ class Engine implements IPlaybackEngine {
     const { duration, currentTime } = this._output;
     let data;
 
-    if (this._adapterStrategy.attachedAdapter) {
-      data = this._adapterStrategy.attachedAdapter.debugInfo;
+    if (this._output.attachedAdapter) {
```
Comment (BrooklynKing) [inline src/modules/playback-engine/playback-engine.ts]: Let's make methods `getDebugInfo` in `output` and move this logic inside it

Comment (maxvektor) [reply src/modules/playback-engine/playback-engine.ts]: done

File: src/modules/playback-engine/output/native/native-otput.ts [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+  constructor({
+    eventEmitter,
+    config,
+    availablePlaybackAdapters = [],
+  }: {
+    eventEmitter: IEventEmitter;
+    config: IPlayerConfig;
+    availablePlaybackAdapters: IPlaybackAdapterClass[];
+  }) {
+    this.on = this.on.bind(this);
```
Comment (BrooklynKing) [inline src/modules/playback-engine/output/native/native-otput.ts]: What is it?

**Comment by BrooklynKing** [comment]
`native-otput.ts` -> `html5video-output`

File: src/modules/playback-engine/playback-engine.ts [hunk 3]
```diff
@@ -93,36 +71,20 @@ class Engine implements IPlaybackEngine {
...
-    return this.attachedAdapter.isSeekAvailable;
+    return this._output.isSeekAvailable;
   }
 
   get isMetadataLoaded() {
-    return this._stateEngine.isMetadataLoaded;
+    return this._output.isMetadataLoaded;
   }
 
   get isPreloadAvailable() {
```
Comment (BrooklynKing) [inline src/modules/playback-engine/playback-engine.ts]: Let's rename is to `isPreloadActive` and move logic to output

Comment (BrooklynKing) [reply src/modules/playback-engine/playback-engine.ts]: Also `isAutoPlayAvailable` -> `isAutoPlayActive` and logic to output

File: src/modules/playback-engine/output/native/html5video-output.ts [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+
+  setLoop(isLoop: boolean) {
+    this._video.loop = isLoop;
+  }
+
+  setPlaybackRate(rate: number) {
+    this._video.playbackRate = rate;
+  }
+
+  setPreload(preload: Preload = 'auto') {
```
Comment (BrooklynKing) [inline src/modules/playback-engine/output/native/html5video-output.ts]: use `PreloadType`

File: src/modules/playback-engine/types.ts [hunk 5]
```diff
@@ -115,28 +116,27 @@ interface IPlaybackEngineDependencies {
...
   destroy: () => void;
 
   getElement: () => HTMLVideoElement | null;
-  getViewDimensions: () => IDimensions;
-
-  on: (event: string, cb: (event?: Event) => void) => void;
-  off: (event: string, cb: (event?: Event) => void) => void;
 
   setPlaybackRate: (rate: number) => void;
   setPreload: (preload: 'auto' | 'metadata' | 'none') => void;
```
Comment (BrooklynKing) [inline src/modules/playback-engine/types.ts]: use `PreloadType`

File: src/modules/playback-engine/output/native/html5video-output.ts [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+import AdapterStrategy from './adapters-strategy';
+import VideoEvent from '../../../../constants/events/video';
+import {
+  isAndroid,
+  isIPad,
+  isIPhone,
+  isIPod,
+} from '../../../../utils/device-detection';
+
+type Preload = 'auto' | 'metadata' | 'none';
```
Comment (BrooklynKing) [inline src/modules/playback-engine/output/native/html5video-output.ts]: use `PreloadType` from `types.ts` in `playback-engine` folder

**Comment by maxvektor** [comment]
@BrooklynKing take a look pls

**Pull request merged**